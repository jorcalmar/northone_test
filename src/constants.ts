import * as httpStatus from 'http-status'

export enum TaskStatuses {
    PENDING = 'PENDING',
    DONE = 'DONE'
}

export enum HttpStatus {
    OK = httpStatus.OK,
    CREATED = httpStatus.CREATED,
    BAD_REQUEST = httpStatus.BAD_REQUEST,
    INTERNAL_SERVER_ERROR = httpStatus.INTERNAL_SERVER_ERROR,
    NOT_FOUND = httpStatus.NOT_FOUND
}