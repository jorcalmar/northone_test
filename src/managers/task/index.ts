import { taskModel, Task } from '../../db/models/task'
import { ITaskInput } from '../../interfaces/task'

export const createTask = async (createTaskInput: ITaskInput): Promise<Task> => {
    const createdTask = await taskModel.create(createTaskInput);

    return createdTask;
}