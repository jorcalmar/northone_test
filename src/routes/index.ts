export * from './home'
export * from './task'
export * from './category'