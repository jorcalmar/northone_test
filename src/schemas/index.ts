export * from './createTaskBodySchema'
export * from './signUpBodySchema'