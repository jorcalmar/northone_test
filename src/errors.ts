import { HttpStatus } from './constants'

export const errors = {
  RESOURCE_NOT_FOUND: {
    statusCode: HttpStatus.NOT_FOUND,
    message: 'Resource was not found',
    name: 'RESOURCE_NOT_FOUND',
  },
  CATEGORY_NOT_FOUND: {
    statusCode: HttpStatus.NOT_FOUND,
    message: 'Category not found',
    name: 'CATEGORY_NOT_FOUND'
  },
  PARENT_NOT_FOUND: {
    statusCode: HttpStatus.NOT_FOUND,
    message: 'Parent not found',
    name: 'PARENT_NOT_FOUND'
  },
  MAX_NESTED_LEVEL_REACHED: {
    statusCode: HttpStatus.BAD_REQUEST,
    message: 'Too many levels of subtasks',
    name: 'MAX_NESTED_LEVEL_REACHED'
  },
  INVALID_DUE_DATE: {
    statusCode: HttpStatus.BAD_REQUEST,
    message: 'Due date is invalid',
    name: 'INVALID_DUE_DATE'
  }
};