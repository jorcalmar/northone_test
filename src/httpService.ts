import { createServer, plugins } from 'restify';
import { dbConnect, closeDb } from './db/connection';
import { homeController, createTaskRoute, updateTaskRoute, deleteTaskRoute, getTasksRoute, createCategoryRoute } from './routes';
import config from 'config';
import { bodySchemaValidatorMiddleware } from './middlewares/bodySchemaValidator';
import { createCategorySchema, createTaskBodySchema, updateTaskBodySchema } from './schemas';

export const app = createServer()

/**
 * Configures and starts service.
 */
export const start = async () => {
    app.use(plugins.bodyParser());

    app.get('/', homeController);

    app.post('/api/v1/task', bodySchemaValidatorMiddleware(createTaskBodySchema), createTaskRoute);
    app.patch('/api/v1/tasks/:taskId', bodySchemaValidatorMiddleware(updateTaskBodySchema), updateTaskRoute);
    app.del('/api/v1/tasks/:taskId', deleteTaskRoute);
    app.get('/api/v1/tasks', getTasksRoute);

    app.post('/api/v1/category', bodySchemaValidatorMiddleware(createCategorySchema), createCategoryRoute);

    await dbConnect(config.get('MONGO_URL'));
    await app.listen(config.get('PORT'));

    console.log(`Service listening`);
}

/**
 * Stops service.
 */
export const close = async () => {
    await closeDb();
    await app.close();
}