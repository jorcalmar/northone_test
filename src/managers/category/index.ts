import { categoryModel, Category } from '../../db/models/category'
import { ICategoryInput } from '../../interfaces/category'

import { errors } from '../../errors'

export const createCategory = (categoryInput: ICategoryInput): Promise<Category> => categoryModel.create(categoryInput)

export const validateCategory = async (categoryId: string): Promise<void> => {
    const foundCategory = await categoryModel.findOne({ id: categoryId })

    if (!foundCategory) throw errors.CATEGORY_NOT_FOUND
}