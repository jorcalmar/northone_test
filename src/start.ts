import * as httpService from './httpService'

process.on('SIGINT', () => {
    console.log('Closing service')
    httpService.close()
})

httpService.start();

