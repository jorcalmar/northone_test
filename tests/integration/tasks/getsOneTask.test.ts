import { serviceHooks } from '../../utils/service/serviceHooks'

import { app } from '../../../src/httpService'
import { createTask } from '../../../src/managers'
import { createTaskInput } from '../../utils/data'
import { HttpStatus } from '../../../src/constants'

import request from 'supertest'

describe('Calls service to get one task', () => {
    serviceHooks()

    const apiRequest = request(app)

    it('Gets tasks successfully', async () => {
        const taskInput = createTaskInput()
        const { id: taskId } = await createTask(taskInput)

        const result = await apiRequest
            .get(`/api/v1/tasks/${taskId}`)
            .expect(HttpStatus.OK)

        expect(result.body.data).toMatchSnapshot()
    })

    it('Gets one task when db is empty', async () => {
        await apiRequest
            .get('/api/v1/tasks/any-id')
            .expect(HttpStatus.NOT_FOUND)
    })

    it('Gets tasks with existing category', async () => {
        const taskInput = createTaskInput()
        const { id: taskId } = await createTask(taskInput)

        const result = await apiRequest
            .get(`/api/v1/tasks/${taskId}`)
            .expect(HttpStatus.OK)

        const returnedTask = result.body.data

        expect(returnedTask).toMatchSnapshot()
    })
})