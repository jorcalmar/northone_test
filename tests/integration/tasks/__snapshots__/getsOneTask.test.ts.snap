// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calls service to get one task Gets one task successfully 1`] = `
Object {
  "category": null,
  "createdAt": Any<String>,
  "deleted": false,
  "description": "description1",
  "dueDate": "2099-01-1",
  "id": Any<String>,
  "status": "PENDING",
  "subTasksIds": Array [],
  "title": "title1",
  "updatedAt": Any<String>,
}
`;

exports[`Calls service to get one task Gets tasks with existing category 1`] = `
Object {
  "category": null,
  "createdAt": Any<String>,
  "deleted": false,
  "description": "description1",
  "dueDate": "2099-01-1",
  "id": Any<String>,
  "status": "PENDING",
  "subTasksIds": Array [],
  "title": "title1",
  "updatedAt": Any<String>,
}
`;
